src/java_bytecode/java_object_factory.cpp

```diff
@@ -110,6 +110,76 @@ class java_object_factoryt
 
 /*******************************************************************\
 
+Function: allocate_dynamic_object
+
+  Inputs:
+    target_expr - expression to which the necessary memory will be allocated
+    allocate_type - type of the object allocated
+    symbol_table - symbol table
+    loc - location in the source
+    output_code - code block to which the necessary code is added
+    cast_needed - Boolean flags saying where we need to cast the malloc site
+
+  Outputs: Expression representing the malloc site allocated.
+
+  Purpose: Generates code for allocating a dynamic object.
+           This is used in allocate_object and for allocating strings
+           in the library preprocessing.
+
+\*******************************************************************/
+
+exprt allocate_dynamic_object(
+  const exprt &target_expr,
+  const typet &allocate_type,
+  symbol_tablet &symbol_table,
+  const source_locationt &loc,
+  code_blockt &output_code,
+  bool cast_needed)
+{
+  // build size expression
+  exprt object_size=size_of_expr(allocate_type, namespacet(symbol_table));
+
+  if(allocate_type.id()!=ID_empty)
+  {
+    assert(!object_size.is_nil() && "size of Java objects should be known");
+    // malloc expression
+    exprt malloc_expr=side_effect_exprt(ID_malloc);
+    malloc_expr.copy_to_operands(object_size);
+    typet result_type=pointer_typet(allocate_type);
+    malloc_expr.type()=result_type;
+    // Create a symbol for the malloc expression so we can initialize
+    // without having to do it potentially through a double-deref, which
+    // breaks the to-SSA phase.
+    symbolt &malloc_sym=new_tmp_symbol(
+      symbol_table,
+      loc,
+      pointer_typet(allocate_type),
+      "malloc_site");
+    code_assignt assign=code_assignt(malloc_sym.symbol_expr(), malloc_expr);
+    code_assignt &malloc_assign=assign;
+    malloc_assign.add_source_location()=loc;
```

Comment:
Probably, but this is not our code (actually it's yours). If you agree I can simplify.